src/storage/src/source/generator/ldbc.rs

```diff
@@ -0,0 +1,230 @@
+                                    if field.is_empty() && typ.nullable {
+                                       return Datum::Null;
+                                    }
+                                  error!("could not decode typ {typ:?} for value '{field}' in ldbc output {}", csv.output);
+                                    panic!("could not decode typ {typ:?} for value '{field}' in ldbc output {}", csv.output);
+                                }
+                            }
+                        },
+                    ));
+                    // Make symmetric.
+                    if csv.output == LDBC_PERSON_KNOWS_PERSON_OUTPUT {
+                        assert!(self.pending.is_none());
+                        let mut iter = row.iter();
+                        let creationdate = iter.next().expect("must exist");
+                        let person1id = iter.next().expect("must exist");
+                        let person2id = iter.next().expect("must exist");
+                        let swapped = Row::pack_slice(&[
+                            creationdate,
+                            person2id,
+                            person1id,
+                        ]);
+                        self.pending = Some((csv.output, swapped));
+                    }
+                    Some((csv.output, row))
+                }
+                Ok(false) => None,
+                Err(_) => None,
+            })
+    }
+}
+
+impl Iterator for Context {
+    type Item = (usize, Event<Option<MzOffset>, (Row, i64)>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((output, row)) = self.pending.take() {
+            self.offset += 1;
+            return Some((output, Event::Message(self.offset, (row, 1))));
+        }
+        loop {
+            match self.next_record() {
+                Some((output, row)) => {
+                    self.offset += 1;
+                    return Some((output, Event::Message(self.offset, (row, 1))));
+                }
+                None => {
+                    // If we're out of records for this file, close the current file and its data.
+                    if let Some(csv) = self.rdr.take() {
+                        return Some((csv.output, Event::Progress(Some(self.offset))));
+                    }
+                    // Otherwise open the next file.
+                    self.next_file();
+                    // No more files.
+                    if self.rdr.is_none() {
+                        return None;
+                    }
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+/// Fetches URLs whose bodies are zstd compressed and returns their response body uncompressed and
+/// concatenated.
+fn fetch_urls_compressed(urls: &[String]) -> Result<Vec<u8>, anyhow::Error> {
+    let mut buf = Vec::new();
+    for url in urls {
+        let resp = reqwest::blocking::get(url)?.error_for_status()?;
```

Comment:
Blocking network requests on a timely thread (if I'm reading this right) are going to be bad for throughput. Probably best to toss this on a Tokio task?

Reply:
See my question in the storage channel: this is in the load generator and nothing nearby is async, so I'm unsure of the best way to do that. Spawning a task won't work because the join handle is returned as a future?

Reply:
Ah, I see. Yeah, sorry, spawning a task is what you'll want to do to avoid timely trying to drive a future, which can result in slow polls and timeouts when networking is involved (see #incident-70). But you'll need _some_ way to stitch that into the operator, and if the `Generator` trait doesn't support either `async` or manual activations, there's no choice but to fix it.

Reply:
FWIW, it doesn't seem terribly hard to make the `Generator` trait `async`? Its only caller is in the `SourceRender` pipeline inside of an `async` block.

Reply:
Made this async.